File: data-stuctures/UnionFind.js
class UnionFind {
  constructor(size) {
    //parent[i] >= 0 - родитель вершины i
    //parent[i] < 0 - если родителя нет, то это минус кол-во вершин в компоненте с корнем в этой вершине        
    this.parent = Array.from({ length: size }, () => -1);
  }

  getRoot(v) {
    let r = v;

    while (this.parent[r] >= 0) {
      r = this.parent[r];
    }

    while (v !== r) {
      const pV = this.parent[v];
      this.parent[v] = r;
      v = pV;
    }

    return r;
  }

  union(u, v) {
    u = this.getRoot(u);
    v = this.getRoot(v);

    if (u === v) {
      return false;
    }

    if (this.parent[u] < this.parent[v]) {
      this.parent[u] += this.parent[v];
      this.parent[v] = u;
    } else {
      this.parent[v] += this.parent[u];
      this.parent[u] = v;
    }

    return true;
  }
}

File: data-stuctures/UnionFind.py
class UnionFind:
    def __init__(self, size: int):
        # parent[i] >= 0 - родитель вершины i
        # parent[i] < 0 - если родителя нет, то это минус кол-во вершин в компоненте с корнем в этой вершине
        self.parent = [-1] * size

    def getRoot(self, v):
        r = v

        while self.parent[r] >= 0:
            r = self.parent[r]

        while v != r:
            pV = self.parent[v]
            self.parent[v] = r
            v = pV

        return r

    def union(self, u, v):
        u = self.getRoot(u)
        v = self.getRoot(v)

        if u == v:
            return False

        if self.parent[u] < self.parent[v]:
            self.parent[u] += self.parent[v]
            self.parent[v] = u
        else:
            self.parent[v] += self.parent[u]
            self.parent[u] = v

        return True
